File: src/RongYokDownloader/Services/JsonExtract.cs
namespace RongYokDownloader.Services;

/// <summary>
/// rongyok.com renders its data as JavaScript literals embedded in the page
/// (e.g. <c>seriesData = [ {...}, ... ]</c>, or an episodes object inside the watch page).
/// These helpers pull a balanced JSON array/object out of the surrounding script text,
/// respecting string literals and escapes so brackets inside Thai descriptions don't confuse it.
/// The returned substring is valid JSON and can be handed straight to System.Text.Json.
/// </summary>
public static class JsonExtract
{
    /// <summary>
    /// Finds <paramref name="marker"/> in <paramref name="source"/>, then returns the balanced
    /// bracketed literal that begins at the next <paramref name="openChar"/> after it.
    /// Returns null if not found or unbalanced.
    /// </summary>
    public static string? BalancedAfter(string source, string marker, char openChar)
    {
        char closeChar = openChar switch { '[' => ']', '{' => '}', _ => throw new ArgumentException("openChar") };

        int at = source.IndexOf(marker, StringComparison.Ordinal);
        if (at < 0) return null;

        int i = at + marker.Length;
        // advance to the first opening bracket
        while (i < source.Length && source[i] != openChar) i++;
        if (i >= source.Length) return null;

        int start = i;
        int depth = 0;
        bool inStr = false, esc = false;
        for (; i < source.Length; i++)
        {
            char ch = source[i];
            if (inStr)
            {
                if (esc) esc = false;
                else if (ch == '\\') esc = true;
                else if (ch == '"') inStr = false;
            }
            else
            {
                if (ch == '"') inStr = true;
                else if (ch == openChar) depth++;
                else if (ch == closeChar)
                {
                    depth--;
                    if (depth == 0) return source.Substring(start, i - start + 1);
                }
            }
        }
        return null; // unbalanced
    }

    /// <summary>Extracts the whole <c>seriesData = [ ... ]</c> catalog array as JSON.</summary>
    public static string? CatalogArray(string html) => BalancedAfter(html, "seriesData", '[');

    /// <summary>Extracts the <c>"episodes":[ ... ]</c> array from a watch page as JSON.</summary>
    public static string? EpisodesArray(string html) => BalancedAfter(html, "\"episodes\"", '[');
}

File: mobile/lib/services/json_extract.dart
/// rongyok.com renders its data as JavaScript literals embedded in the page
/// (e.g. `seriesData = [ {...}, ... ]`, or an `"episodes":[...]` array inside the
/// watch page). These helpers pull a *balanced* bracketed literal out of the
/// surrounding script text, respecting string literals and escapes so brackets
/// inside Thai descriptions don't confuse it.
///
/// Direct port of RongYokDownloader.Services.JsonExtract (C#). The returned
/// substring is valid JSON and can be handed straight to `jsonDecode`.
class JsonExtract {
  /// Finds [marker] in [source], then returns the balanced bracketed literal
  /// that begins at the next [openChar] after it. Returns null if not found or
  /// unbalanced.
  static String? balancedAfter(String source, String marker, String openChar) {
    final String closeChar = switch (openChar) {
      '[' => ']',
      '{' => '}',
      _ => throw ArgumentError.value(openChar, 'openChar'),
    };

    final at = source.indexOf(marker);
    if (at < 0) return null;

    var i = at + marker.length;
    // advance to the first opening bracket
    while (i < source.length && source[i] != openChar) {
      i++;
    }
    if (i >= source.length) return null;

    final start = i;
    var depth = 0;
    var inStr = false;
    var esc = false;
    for (; i < source.length; i++) {
      final ch = source[i];
      if (inStr) {
        if (esc) {
          esc = false;
        } else if (ch == r'\') {
          esc = true;
        } else if (ch == '"') {
          inStr = false;
        }
      } else {
        if (ch == '"') {
          inStr = true;
        } else if (ch == openChar) {
          depth++;
        } else if (ch == closeChar) {
          depth--;
          if (depth == 0) return source.substring(start, i + 1);
        }
      }
    }
    return null; // unbalanced
  }

  /// Extracts the whole `seriesData = [ ... ]` catalog array as JSON.
  static String? catalogArray(String html) => balancedAfter(html, 'seriesData', '[');

  /// Extracts the `"episodes":[ ... ]` array from a watch page as JSON.
  static String? episodesArray(String html) => balancedAfter(html, '"episodes"', '[');
}
